src/poprox_recommender/components/diversifiers/locality_calibration.py

```diff
@@ -152,3 +166,145 @@ def compute_topic_prefs(self, interest_profile):
+    return weight
+
+
+def related_indices(
+    selected_subhead: str, selected_date: datetime, clicked_articles: list, time_decay: bool, topk_similar: int
+):
+    all_subheads = [selected_subhead] + [article.subhead for article in clicked_articles] 
+    #is_original = [1 if article.is_original == 1 else 0 for article in clicked_articles]
+
+    embeddings = model.encode(all_subheads)
+
+    target_embedding = embeddings[0].reshape(1, -1)
+    clicked_embeddings = embeddings[1:]
+    similarities = cosine_similarity(target_embedding, clicked_embeddings)[0]
+    #similarities = [similarities[i] * is_original[i] for i in range(len(is_original))]
+
+    if time_decay:
+        weights = [
+            get_time_weight(selected_date, published_date)
+            for published_date in [article.published_at for article in clicked_articles]
+        ]
+        similarities = similarities * weights
+
+    # check if has similarity above the threshold 
+    if np.max(similarities) < SEMANTIC_THRESHOLD:
+        # If no articles are similar enough, 
+        # we'll default to a general appraoch that 
+        # uses the topk similar articles to generate a broad topic inspired context for the rec?
+        return np.argsort(similarities)[-topk_similar:][::-1]
+    else: 
+        return np.argsort(similarities)[-1:][::-1]
+
+    
+
+
+def related_context(
+    article: Article, clicked_articles: ArticleSet, time_decay: bool, topk_similar: int
+):
+    selected_subhead = article.subhead
+    selected_date = article.published_at
+
+    clicked_articles = clicked_articles.articles
+
+    candidate_indices = related_indices(selected_subhead, selected_date, clicked_articles, time_decay, topk_similar)
+
+    return [clicked_articles[index] for index in candidate_indices]
+
+
+def generated_context(
+    article: Article, clicked_articles: ArticleSet, time_decay: bool, topk_similar: int
+):
+
+    topk_similar = min(topk_similar, len(clicked_articles.articles))
+    related_articles = related_context(article, clicked_articles, time_decay, topk_similar)
+
+    if len(related_articles) == 1: 
+        # high similarity, use the article to rewrite the subhead 
+
+        news_list = {'MAIN NEWS': article.subhead , 
+                     'RELATED NEWS': related_articles[0].subhead}
+        
+        input_prompt = f"{news_list}"
+
+        generated_subhead = semantic_narrative(input_prompt)
+
+    else: 
+        news_list = {'MAIN NEWS': article.subhead, 
+                     'RELATED NEWS': [related_articles[i].subhead for i in range(topk_similar)]}
+        
+        input_prompt = f"{news_list}"
+        generated_subhead = highlevel_narrative(input_prompt)
+
+    return generated_subhead
+
+
+def semantic_narrative(news_list):
+    system_prompt = ("You are an expert to rewrite the subheadline of MAIN NEWS in a natural and factural tone. "
+    "I will provide you with two news articles: MAIN NEWS and RELATED NEWS that occurred before the MAIN NEWS. "
```

Comment:
Will this bring a little bit confusion? Sounds like it can also be interpreted as the RELATED NEWS is placed before MAIN NEWS in the provided context.
I'm thinking if we shall mention that related news is something user has clicked before. 
`MAIN NEWS to be recommended and RELATED NEWS that the user clicked before`.